lib/midi-smtp-server/tls-transport.rb

```diff
@@ -43,7 +43,15 @@ def initialize(cert_path, key_path, ciphers, methods, cert_cn, cert_san, logger)
         raise "File \”#{@cert_path}\" does not exist or is not a regular file. Could not load certificate." unless File.file?(@cert_path.to_s)
         raise "File \”#{@key_path}\" does not exist or is not a regular file. Could not load private key." unless File.file?(@key_path.to_s)
         # try to load certificate and key
-        @ssl_context.cert = OpenSSL::X509::Certificate.new(File.open(@cert_path.to_s))
+        cert_lines = File.read(@cert_path.to_s).lines
+        cert_indexes = cert_lines.each_with_index.map { |line, index| index if line.downcase.include?('begin cert') }.compact
+        certs = []
+        cert_indexes.each_with_index do |cert_index, current_index|
+          end_index = (cert_indexes[current_index + 1] || 0) - 1
+          certs << OpenSSL::X509::Certificate.new(cert_lines[cert_index..end_index].join)
+        end
+        @ssl_context.cert = certs[0]
```

Comment:
@ocha Luri, in a chained PEM file isn't always the `latest` cert the server cert?

It might be better to do it as:

```ruby
@ssl_context.cert = certs.last
@ssl_context.chain = certs
```

What do you think?

Reply:
@TomFreudenberg I believe root certificate is last and actual server cert is first. That is how Let's Encrypt certbot constructs the fullchain.pem file. If digicert can be trusted, here is their guide https://www.digicert.com/kb/ssl-support/pem-ssl-creation.htm

Reply:
Well you are right, I will check and make a small rework after merging your PR so that it will be able to contain also the KEY file in the PEM file.

Checkout last section `Creating a .pem with the Private Key and Entire Trust Chain` from your Digicert link.

Reply:
Should be:

```ruby
@ssl_context.cert = certs.first
@ssl_context.chain = certs
```

Reply:
Yes. I saw that part. I was surprised when I read it since I have never seen private key bundled with cert before - you learn something new every day!

Reply:
Yeah - but you find similar in the OVPN files for example - so it seems to happen

Reply:
I was going through ruby docs and it [seems like you could just give full chain PEM file before and openssl lib would figure it out for you](https://ruby-doc.org/stdlib-2.6.1/libdoc/openssl/rdoc/OpenSSL/SSL/SSLContext.html#method-i-add_certificate_chain_file). Wonder why they removed it

Reply:
I also tried using `#add_certificate(certiticate, pkey [, extra_certs])` as documentation recommends, but it would not set `cert` and `key` variables on `ssl_context`, so I reverted back to the "old way"™️

Reply:
I found this PR merged in openssl - this looks like what we are looking for?

Have you seen that?

https://github.com/ruby/openssl/pull/441

Reply:
I haven't seen that PR. That would perfect if we can use it. I was hoping SSL lib would handle parsing and loading chain file. It felt wrong to implement that logic here, but did not find anything in ruby docs. Should we add a Todo line to switch to that method you linked once it is available?

Reply:
Yes - I append a note and an issue later.

For now I will merge your changes as all fits.

If it is ok for you, I will follow up with some checks for tests and updates before pushing this in a new release.

Reply:
@TomFreudenberg That's totally fine. Thank you 🙏